### PR TITLE
Make `CupertinoRadio`'s `mouseCursor` a `WidgetStateProperty`

### PR DESCRIPTION
**Pull request by victorsanni**
https://github.com/flutter/flutter/pull/149681 introduced `mouseCursor `to `CupertinoRadio` as a `MouseCursor` instead of a `WidgetStateProperty` to match Material Radio's `mouseCursor` property for `.adaptive`.

This PR changes `mouseCursor` to be of type `WidgetStateProperty<MouseCursor>` as per review comments in https://github.com/flutter/flutter/pull/151788#discussion_r1680538286.

PR bringing `mouseCursor` into `CupertinoRadio`: https://github.com/flutter/flutter/pull/149681.

Part of https://github.com/flutter/flutter/issues/58192

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [ ] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#overview
[Tree Hygiene]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md
[test-exempt]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md
[Features we expect every widget to implement]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/blob/main/docs/contributing/Chat.md
[Data Driven Fixes]: https://github.com/flutter/flutter/blob/main/docs/contributing/Data-driven-Fixes.md

### CODE REVIEW

**Comment by victorsanni** [comment]
cc @jonahwilliams

File: packages/flutter/lib/src/cupertino/radio.dart [hunk 1]
```diff
@@ -297,7 +294,7 @@ class _CupertinoRadioState<T> extends State<CupertinoRadio<T>> with TickerProvid
       checked: widget._selected,
       selected: accessibilitySelected,
       child: buildToggleable(
-        mouseCursor: effectiveMouseCursor,
+        mouseCursor: widget.mouseCursor ?? _defaultMouseCursor,
```
Comment (justinmc) [inline packages/flutter/lib/src/cupertino/radio.dart:297]: Can you set this as the actual default value for CupertinoRadio.mosueCursor or is it more complicated than that?

File: packages/flutter/lib/src/material/radio.dart [hunk 2]
```diff
@@ -447,7 +447,9 @@ class _RadioState<T> extends State<Radio<T>> with TickerProviderStateMixin, Togg
               value: widget.value,
               groupValue: widget.groupValue,
               onChanged: widget.onChanged,
-              mouseCursor: widget.mouseCursor,
+              mouseCursor: WidgetStateProperty.all(widget.mouseCursor
+                ?? (kIsWeb && widget.onChanged != null
+                  ? SystemMouseCursors.click : SystemMouseCursors.basic)),
```
Comment (justinmc) [inline packages/flutter/lib/src/material/radio.dart:452]: Is this duplicating CupertinoRadio._defaultMouseCursor? Could you make that public and use it here?

File: packages/flutter/test/cupertino/radio_test.dart [hunk 3]
```diff
@@ -463,13 +463,25 @@ void main() {
     final FocusNode focusNode = FocusNode(debugLabel: 'Radio');
     tester.binding.focusManager.highlightStrategy = FocusHighlightStrategy.alwaysTraditional;
 
+    MouseCursor getMouseCursor(Set<WidgetState> states) {
+      if (states.contains(WidgetState.disabled)) {
+        return SystemMouseCursors.forbidden;
+      }
+      if (states.contains(WidgetState.focused)){
```
Comment (justinmc) [inline packages/flutter/test/cupertino/radio_test.dart:470]: Missing a space: `) {`.

**Review by justinmc — COMMENTED** [review]
Some quick questions, otherwise looks good.

File: packages/flutter/lib/src/material/radio.dart (hunk 2 above)
Comment (victorsanni) [reply packages/flutter/lib/src/material/radio.dart]: I added a static method `defaultMouseCursor` in `CupertinoRadio` that I just call here to get the default mouse cursor.

File: packages/flutter/lib/src/cupertino/radio.dart [hunk 4]
```diff
@@ -132,24 +132,22 @@ class CupertinoRadio<T> extends StatefulWidget {
...
-  ///  * [WidgetState.hovered].
   ///  * [WidgetState.focused].
   ///  * [WidgetState.disabled].
   ///
-  /// If null, then [SystemMouseCursors.basic] is used when this radio button is disabled.
-  /// When this radio button is enabled, [SystemMouseCursors.click] is used on Web, and
-  /// [SystemMouseCursors.basic] is used on other platforms.
+  /// If null, then [SystemMouseCursors.basic] is used when this radio button is
+  /// disabled. When this radio button is enabled, [SystemMouseCursors.click] is
+  /// used on Web, and [SystemMouseCursors.basic] is used on other platforms.
```
Comment (justinmc) [inline packages/flutter/lib/src/cupertino/radio.dart:143]: Nit: You might just say `defaults to [defaultMouseCursor]` and let people look that up themselves. Or maybe it's more useful for users to see it written out like this, up to you.

File: packages/flutter/lib/src/cupertino/radio.dart [hunk 5]
```diff
@@ -132,24 +132,22 @@ class CupertinoRadio<T> extends StatefulWidget {
...
   ///
   /// See also:
   ///
   ///  * [WidgetStateMouseCursor], a [MouseCursor] that implements
   ///    `WidgetStateProperty` which is used in APIs that need to accept
-  ///    either a [MouseCursor] or a [WidgetStateProperty<MouseCursor>].
-  final MouseCursor? mouseCursor;
+  ///    either a [MouseCursor] or a [WidgetStateProperty].
```
Comment (justinmc) [inline packages/flutter/lib/src/cupertino/radio.dart:149]: Why did you remove the type annotation on WidgetStateProperty?

File: packages/flutter/lib/src/cupertino/radio.dart [hunk 6]
```diff
@@ -210,6 +208,13 @@ class CupertinoRadio<T> extends StatefulWidget {
 
   bool get _selected => value == groupValue;
 
+  /// The default mouse cursor of a [CupertinoRadio].
```
Comment (justinmc) [inline packages/flutter/lib/src/cupertino/radio.dart:211]: Nit: "mouse cursor" => "[mouseCursor]"

File: packages/flutter/lib/src/cupertino/radio.dart [hunk 7]
```diff
@@ -132,24 +132,22 @@ class CupertinoRadio<T> extends StatefulWidget {
...
+  /// used on Web, and [SystemMouseCursors.basic] is used on other platforms.
   ///
   /// See also:
   ///
   ///  * [WidgetStateMouseCursor], a [MouseCursor] that implements
   ///    `WidgetStateProperty` which is used in APIs that need to accept
-  ///    either a [MouseCursor] or a [WidgetStateProperty<MouseCursor>].
-  final MouseCursor? mouseCursor;
+  ///    either a [MouseCursor] or a [WidgetStateProperty].
+  final WidgetStateProperty<MouseCursor>? mouseCursor;
```
Comment (justinmc) [inline packages/flutter/lib/src/cupertino/radio.dart:150]: What happens if you explicitly pass `null` here?

**Comment by Piinks** [comment]
Just sharing here, we synced offline that the original change was cut into the beta release that is about to go out. This should be CP'd into beta after it lands. 👍

File: packages/flutter/lib/src/cupertino/radio.dart [hunk 8]
```diff
@@ -210,6 +208,13 @@ class CupertinoRadio<T> extends StatefulWidget {
 
   bool get _selected => value == groupValue;
 
+  /// The default mouse cursor of a [CupertinoRadio].
+  static MouseCursor defaultMouseCursor(Function? onChanged) {
```
Comment (justinmc) [inline packages/flutter/lib/src/cupertino/radio.dart:212]: Could this be an actual `WidgetStateProperty<MouseCursor>` instead of a function that returns a MouseCursor?

File: packages/flutter/lib/src/cupertino/radio.dart [hunk 9]
```diff
@@ -297,7 +293,7 @@ class _CupertinoRadioState<T> extends State<CupertinoRadio<T>> with TickerProvid
       checked: widget._selected,
       selected: accessibilitySelected,
       child: buildToggleable(
-        mouseCursor: effectiveMouseCursor,
+        mouseCursor: widget.mouseCursor ?? WidgetStateProperty.all(CupertinoRadio.defaultMouseCursor(widget.onChanged)),
```
Comment (justinmc) [inline packages/flutter/lib/src/cupertino/radio.dart:296]: Instead of doing this here, can you set the default in the parameter list of the constructor?

File: packages/flutter/lib/src/material/radio.dart [hunk 10]
```diff
@@ -447,7 +447,8 @@ class _RadioState<T> extends State<Radio<T>> with TickerProviderStateMixin, Togg
               value: widget.value,
               groupValue: widget.groupValue,
               onChanged: widget.onChanged,
-              mouseCursor: widget.mouseCursor,
+              mouseCursor: WidgetStateProperty.all(widget.mouseCursor
+                ?? CupertinoRadio.defaultMouseCursor(widget.onChanged)),
```
Comment (justinmc) [inline packages/flutter/lib/src/material/radio.dart:451]: Same here, can you do this in the constructor parameter list?

**Review by justinmc — COMMENTED** [review]
Some quick questions.

File: packages/flutter/lib/src/cupertino/radio.dart (hunk 4 above)
Comment (victorsanni) [reply packages/flutter/lib/src/cupertino/radio.dart:143]: True. Then I could move these docs to `defaultMouseCursor`.

File: packages/flutter/lib/src/cupertino/radio.dart (hunk 5 above)
Comment (victorsanni) [reply packages/flutter/lib/src/cupertino/radio.dart:149]: I think @Piinks mentioned previously that it doesn't work with dartdocs? I might be wrong though.

File: packages/flutter/lib/src/cupertino/radio.dart (hunk 7 above)
Comment (victorsanni) [reply packages/flutter/lib/src/cupertino/radio.dart:150]: I think the null coalescing operator should make it resolve to `WidgetStateProperty.all(CupertinoRadio.defaultMouseCursor(widget.onChanged))`.

File: packages/flutter/lib/src/cupertino/radio.dart (hunk 8 above)
Comment (victorsanni) [reply packages/flutter/lib/src/cupertino/radio.dart:212]: The `mouseCursor` property of Material `Radio` is nullable, so to pass it to `WidgetStateProperty<MouseCursor>?` which is the type of `CupertinoRadio`'s `mouseCursor` it would need to have resolved to a non-null `MouseCursor` which we can then wrap with `WidgetStateProperty.all`. If the `defaultMouseCursor` function returns a `WidgetStateProperty<MouseCursor>`, there would be an extra step to resolve that into a plain non-null `MouseCursor` which would then be wrapped with `WidgetStateProperty.all`.

There might be a better way to do all this though that I am missing.

File: packages/flutter/lib/src/cupertino/radio.dart (hunk 9 above)
Comment (victorsanni) [reply packages/flutter/lib/src/cupertino/radio.dart:296]: `defaultMouseCursor` takes `widget.onChanged` as an argument. Since `widget.onChanged` is non-`const`, `defaultMouseCursor` can't be set as the default `mouseCursor` in the parameter list.

File: packages/flutter/lib/src/cupertino/radio.dart (hunk 5 above)
Comment (justinmc) [reply packages/flutter/lib/src/cupertino/radio.dart:147]: Ah that's probably right :+1:

File: packages/flutter/lib/src/material/radio.dart [hunk 11]
```diff
@@ -447,7 +447,10 @@ class _RadioState<T> extends State<Radio<T>> with TickerProviderStateMixin, Togg
               value: widget.value,
               groupValue: widget.groupValue,
               onChanged: widget.onChanged,
-              mouseCursor: widget.mouseCursor,
+              mouseCursor: WidgetStateProperty.resolveWith((Set<MaterialState> states) {
+                return MaterialStateProperty.resolveAs<MouseCursor?>(widget.mouseCursor, states)
+                  ?? CupertinoRadio.defaultMouseCursor(widget.onChanged);
+              }),
```
Comment (justinmc) [inline packages/flutter/lib/src/material/radio.dart:453]: If CupertinoRadio.defaultMouseCursor returns a WidgetStateProperty then could you do this instead?

```dart
              mouseCursor: widget.mouseCursor == null
                ? CupertinoRadio.defaultMouseCursor(widget.onChanged)
                : WidgetStateProperty.all<MouseCursor>(widget.mouseCursor!),
```

File: packages/flutter/lib/src/cupertino/radio.dart (hunk 8 above)
Comment (justinmc) [reply packages/flutter/lib/src/cupertino/radio.dart:214]: If I'm understanding correctly, try this below and then see my other comment about Material Radio using this.

```dart
  static WidgetStateProperty<MouseCursor> defaultMouseCursor(Function? onChanged) {
    final MouseCursor mouseCursor = (onChanged != null && kIsWeb)
      ? SystemMouseCursors.click
      : SystemMouseCursors.basic;
    return WidgetStateProperty.all<MouseCursor>(mouseCursor);
  }
```

**Review by justinmc — APPROVED** [review]
Some ideas about how to make defaultMouseCursor be a WidgetStateProperty, but LGTM :+1: either way.

File: packages/flutter/lib/src/material/radio.dart (hunk 11 above)
Comment (victorsanni) [reply packages/flutter/lib/src/material/radio.dart]: This works, I forgot we could just use `!` to keep the analyzer quiet. Thanks for the insight!

**Comment by victorsanni** [comment]
@Piinks should this PR be reverted until final discussions on `WidgetStateProperty`? Or should the `mouseCursor` property be removed entirely (and then cherrypick that removal into beta)?

**Comment by Piinks** [comment]
Let's revert this, and leave the mouseCursor property that made into beta as is. 👍

**Comment by victorsanni** [comment]
Reason for revert: Needs further discussion on the pros and cons of adding new properties as `WidgetStateProperty`
